flink-runtime/src/main/java/org/apache/flink/runtime/rpc/akka/AkkaInvocationHandler.java

```diff
@@ -205,12 +207,34 @@ private Object invokeRpc(Method method, Object[] args) throws Exception {
 			result = null;
 		} else if (Objects.equals(returnType, CompletableFuture.class)) {
 			// execute an asynchronous call
-			result = ask(rpcInvocation, futureTimeout);
+			CompletableFuture resultFuture = ask(rpcInvocation, futureTimeout);
+
+			result = resultFuture.thenApply((Object o) -> {
+				if (o instanceof SerializedValue) {
+					try {
+						return  ((SerializedValue) o).deserializeValue(getClass().getClassLoader());
+					} catch (IOException | ClassNotFoundException e) {
+						throw new CompletionException(e);
+					}
+				} else {
+					return o;
+				}
+			});
 		} else {
 			// execute a synchronous call
-			CompletableFuture<?> futureResult = ask(rpcInvocation, futureTimeout);
+			CompletableFuture<?> resultFuture = ask(rpcInvocation, futureTimeout);
 
-			result = futureResult.get(futureTimeout.getSize(), futureTimeout.getUnit());
+			Object returnedResult = resultFuture.get(futureTimeout.getSize(), futureTimeout.getUnit());
+
+			if (returnedResult instanceof SerializedValue) {
+				try {
+					result = ((SerializedValue) returnedResult).deserializeValue(getClass().getClassLoader());
+				} catch (IOException | ClassNotFoundException e) {
+					throw new CompletionException(e);
+				}
+			} else {
+				result = returnedResult;
+			}
```

Comment:
This is somewhat a code duplication with the other branch. Could we deduplicate this code by applying the deserialization logic on the future? Additionally it would be good to catch `ExecutionException` when calling `resultFuture.get()` and unstrip them via `ExceptionUtils.unstripExecutionException`.

Reply:
I did not find `ExceptionUtils#unstripExecutionException`, but exists `ExceptionUtils#stripExecutionException`. Did you mean `ExceptionUtils#rethrow`?

Reply:
Hi @tillrohrmann I have updated the PR based on your suggestion, except this one. I did not find the method you mentioned.